### ISSUE TITLE
Be explicit about the format of all “text” fields

### ISSUE THREAD

**Issue by kpreid** [gnuradio/SigMF#7]
I was asked to review the SigMF specification by @bhilburn. This is my first detailed writeup of an issue.

There are several fields in the metadata definition which are said to be “text” or “human-readable”. I recommend that it be specified explicitly and exactly what sort of text. Software and data authors may otherwise make incompatible interpretations, consciously or accidentally, such as:

1. The text is parsed as HTML or some subset thereof.
2. The text is plain text but whitespace is not significant. (That is, the result of putting the text through a HTML entity escaper but not processing whitespace.)
3. The text is plain text and newlines are significant but multiple spaces are not (as in Markdown, loosely speaking).
4. The text is plain text and all whitespace is significant (as in `<pre>`).

(The specific examples here are mostly about displaying SigMF metadata in web browsers, but this will presumably be a very common application anyway.)

I would recommend choosing either interpretation 1 or interpretation 4. Interpretation 4 has the advantage of simplicity; Interpretation 1 has the advantage of being able to include hyperlinks to additional documentation, license documents, authors' sites, catalogs of related files, and so on.

The currently-defined fields for which this clarification should be made are global `description`, `author`, `license`, `hw` and annotation `comment`.

**Comment by smunaut** [comment]
I'd go with 4.
Having to parse HTML in things like CLI programs displaying infos about files would be a MASSIVE pain.

**Comment by kpreid** [comment]
Not necessarily as much of a pain as you might think; for example, a full-featured already-available method would be to pipe the HTML through `lynx -dump -stdin`, with some postprocessing of the whitespace for probably-single-line things like `comment`. Lighter weight solutions could be provided as libraries for various languages or reimplemented as convenient — "apply existing HTML parser, insert blank lines at `<p>` and `<br>`, ignore everything else" should not be particularly verbose.

```
$ echo '<p><a href="https://github.com">Hello.</a><p>Goodbye.' | time lynx -dump -stdin 
   [1]Hello.

   Goodbye.

References

   1. https://github.com/
        0.01 real         0.00 user         0.00 sys

```
That said, I'm not particularly attached to the HTML idea (other than that I think making it possible to have hyperlinks is a good idea and inventing a nonstandard format for them would be worse); my main recommendation here is just that the specification should specify the choice.

**Comment by bhilburn** [comment]
@kpreid - This is a really great point, and something we hadn't considered when first drafting the spec. Thanks for raising it!

The ability to include some form of mark-up in the text fields would be really useful, and would expand the capabilities of things like text annotations in what I think is a really useful way. Sylvain has a good point, though, in that if we mandate the ability to interpret mark-up, it complicates what is necessary to create a `SigMF Compliant` application. If the application doesn't already have an HTML parsing engine in it, for example, requiring one to be able to understand text fields within SigMF is something of a pain.

I know the point of your issue, @kpreid, wasn't necessarily to say that we should or shouldn't support mark-up, but I really like the enabling it, and at the very least, I don't want to preclude it.

That said, I can't think of a way to bring mark-up support into `core` without making things more difficult. Perhaps the answer is we explicitly state that all text fields are plaintext and whitespace is to be respected. We can then allow "turning on" mark-up through an extension.

Thoughts?

**Comment by kpreid** [comment]
> We can then allow "turning on" mark-up through an extension.

I strongly recommend against the pattern of “the presence of this additional field reinterprets existing fields”, it being itself a possible source of misinterpretations.

A more robust alternative would be to at some point define an extension namespace which parallels the core namespace, so that you have (names only for illustration, not suggested!) `core:description` and `withhtml:description`. This is also backwards-compatible in that a document can contain both versions.

**Comment by bhilburn** [comment]
Excellent point, @kpreid! Will also update #13 with that feedback.

**Comment by bhilburn** [comment]
Okay, so there are two parts of this:

1) Specifying the format of "text" fields
2) Enabling HTML markup.

For (1), see #53. For (2), that will be included in the work in #13.

If you have comments on this approach, please discuss here. If you have comments on (1) or (2), please comment in #53 or #13, respectively.
